pkg/sql/schemachanger/scexec/scmutationexec/dependencies.go

```diff
@@ -40,42 +33,54 @@ type Clock interface {
-// schema change had already completed.
-type SyntheticDescriptorStateUpdater interface {
-
-	// AddSyntheticDescriptor sets a synthetic descriptor to shadow any existing
-	// descriptor with the same name or the same ID for the remainder of the
-	// current transaction.
-	AddSyntheticDescriptor(desc catalog.Descriptor)
+// NameResolver is used to retrieve fully qualified names from the catalog.
+type NameResolver interface {
+
+	// GetFullyQualifiedName gets the fully qualified name from a descriptor ID.
+	GetFullyQualifiedName(ctx context.Context, id descpb.ID) (string, error)
 }
 
-// MutationVisitorStateUpdater is the interface for updating the visitor state.
-type MutationVisitorStateUpdater interface {
+// DescriptorReader is used to retrieve descriptors from the catalog.
+type DescriptorReader interface {
+
+	// MustReadImmutableDescriptors reads descriptors from the catalog by ID.
+	MustReadImmutableDescriptors(ctx context.Context, ids ...descpb.ID) ([]catalog.Descriptor, error)
+
+	// MustReadMutableDescriptor the mutable equivalent to
+	// MustReadImmutableDescriptors.
+	MustReadMutableDescriptor(ctx context.Context, id descpb.ID) (catalog.MutableDescriptor, error)
+}
 
-	// GetDescriptor returns a checked-out descriptor, or reads a descriptor from
-	// the catalog by ID if it hasn't been checked out yet.
-	GetDescriptor(ctx context.Context, id descpb.ID) (catalog.Descriptor, error)
+// ImmediateMutationStateUpdater contains the methods used to update the
+// set of changes to bring about from executing immediate mutations.
+type ImmediateMutationStateUpdater interface {
 
-	// CheckOutDescriptor reads a descriptor from the catalog by ID and marks it
-	// as undergoing a change.
-	CheckOutDescriptor(ctx context.Context, id descpb.ID) (catalog.MutableDescriptor, error)
+	// AddToCheckedOutDescriptors adds a mutable descriptor to the set of
+	// checked-out descriptors.
+	AddToCheckedOutDescriptors(mut catalog.MutableDescriptor)
 
-	// AddDrainedName marks a namespace entry as being drained.
-	AddDrainedName(id descpb.ID, nameInfo descpb.NameInfo)
+	// MaybeGetCheckedOutDescriptor looks up a checked-out descriptor by ID.
+	MaybeGetCheckedOutDescriptor(id descpb.ID) catalog.MutableDescriptor
+
+	// DeleteName marks a namespace entry as being drained.
+	DeleteName(id descpb.ID, nameInfo descpb.NameInfo)
 
 	// DeleteDescriptor adds a descriptor for deletion.
 	DeleteDescriptor(id descpb.ID)
 
-	// DeleteComment removes comments for a descriptor
+	// DeleteComment removes comments for a descriptor.
 	DeleteComment(id descpb.ID, subID int, commentType catalogkeys.CommentType)
 
-	// AddComment adds comments for a descriptor
+	// AddComment adds comments for a descriptor.
 	AddComment(id descpb.ID, subID int, commentType catalogkeys.CommentType, comment string)
 
+	// WithReset provisions resetting the in-txn catalog state
+	// to undo the modifications from  earlier stages.
+	WithReset()
```

Comment:
nit: why not just `Reset`? Is it trying to imply that only after this stage ends, the state will be reset? `EnqueueReset`? I don't feel strongly. I just thing of `With` as being used either for functional options or for functions which return a new version of the interface with the new property. Feel free to ignore this one.

Reply:
It's not `Reset` because of a name collision at some point. I'll pick a better name, you're right it's bad.